src/typechecker.erl

```diff
@@ -1933,11 +1939,11 @@ do_type_check_expr_in(Env, Ty, Atom = {atom, LINE, _}) ->
             throw({type_error, Atom, LINE, Ty})
     end;
 do_type_check_expr_in(Env, Ty, Char = {char, _, _}) ->
-    case subtype(type(char), Ty, Env#env.tenv) of
+    case subtype(Char, Ty, Env#env.tenv) of
```

Comment:
this is the main fix: although the whole `char()` range might not be a subtype of `Ty`, but the singleton `Char` type can still be.